### ISSUE TITLE
Nice to have  more types in  DataFrameSuiteBaseLike.approxEquals

### ISSUE THREAD

**Issue by matmsh** [holdenk/spark-testing-base#73]
At present the following test involving an Array will fail.

   val arr1 = Array("a","b")
    val arr2 = Array("a","b")
    val r1 = Row(arr1)
    val r2 = Row(arr2)

    assert(DataFrameSuiteBase.approxEquals(r1,r2, 0.0))

This is because  the comparsion would use arr1 != arr2 , which is true.
The comparison should use  !arr1.sameElements(arr2)

Maybe  DataFrameSuiteBaseLike.approxEquals should also cater for Array[T], where T is String, Float, Long, Double. 

Shing

**Comment by holdenk** [comment]
Thats a good point, we should audit the supported types and make sure we aren't miss handling anything else as well.

**Comment by mahmoudhanafy** [comment]
It's not even supported in Spark, if you call:
`r1.equals(r2)` This will also return false !!

Spark SQL supports some [Data Types](http://spark.apache.org/docs/latest/sql-programming-guide.html#data-types)
As  mentioned in the documentation, If you want to use `ArrayType` you have to use `scala.collection.Seq`

So we can rewrite your example according to the documentation
```Scala
val arr1 = Seq("a","b")
val arr2 = Seq("a","b")
val r1 = Row(arr1)
val r2 = Row(arr2)

r1.equals(r2) // this returns true
```

**Comment by matmsh** [comment]
Silly me !  I should have read the documentation !

One column in  my data frame  is an  array of binaries:
StructField("ips",ArrayType(BinaryType),nullable = false)

In Scala, it will be Seq[Array[Byte]]. 
DataFrameSuiteBase.approxEquals  does not compare  Seq[Array[Byte]] correctly
(  because it will be  handled by
       case _ =>
              if (o1 != o2) return false
)

The following test fails. 
```
  val ba1 = Seq( Array(1.toByte, 2.toByte), Array(3.toByte,4.toByte))
  val ba2 =  Seq( Array(1.toByte, 2.toByte), Array(3.toByte,4.toByte))
  val r1 = Row(ba1)
   val r2 = Row(ba2)

  assert(DataFrameSuiteBase.approxEquals(r1,r2, 0.0)
```

A fix for the above would be to add the following case to  DataFrameSuiteBase.approxEquals.
```
case seqBin1 : Seq[Array[Byte]] =>

                val seqBin2 = o2.asInstanceOf[Seq[Array[Byte]]]

                if (seqBin1.size != seqBin2.size) {
                   return  false
                } else {
                  val eq = seqBin1.zip(seqBin2).forall {
                    case (b1, b2) => java.util.Arrays.equals(b1, b2)
                  }
                  if (!eq) return false

                }

```

It would be nice to have DataFrameSuiteBase.approxEquals working for 
StructField("ips",ArrayType(BinaryType),nullable = false)


Thanks for providing a useful testing framework  !

Shing

**Comment by mahmoudhanafy** [comment]
You are right, I will add this type to DataFrameSuiteBase, But I think this fix has to be added to Spark.
